encoding: define EncodeTableKey / DecodeTableKey for Geo types

otan opened the pull request:
For GROUP BY to work, there needs to be a way of encoding / decoding
table keys for each column type. As such, define EncodeTableKey and
DecodeTableKey for geo types to be their underlying protobuf
representation (we currently have no better way!). Note that this
does not yet allow us to support PKs for geospatial (we could, but we
need some more iteration beforehand).

I've shifted the array descriptor markers around as they seem to be new
in 20.2 as well -- so they should be stable in the upcoming alpha
release.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49711)
<!-- Reviewable:end -->

otan committed:
encoding: define EncodeTableKey / DecodeTableKey for Geo types

For GROUP BY to work, there needs to be a way of encoding / decoding
table keys for each column type. As such, define EncodeTableKey and
DecodeTableKey for geo types to be their underlying protobuf
representation (we currently have no better way!). Note that this
does not yet allow us to support PKs for geospatial (we could, but we
need some more iteration beforehand).

I've shifted the array descriptor markers around as they seem to be new
in 20.2 as well -- so they should be stable in the upcoming alpha
release.

Release note: None

rohany commented on pkg/util/encoding/encoding.go:
```diff
@@ -81,8 +81,9 @@ const (
-	timeTZMarker = bitArrayDescMarker + 1
-	geoMarker    = timeTZMarker + 1
+	timeTZMarker  = bitArrayDescMarker + 1
+	geoMarker     = timeTZMarker + 1
```
I'm not sure why you need this -- it looks like you're just serializing geos to bytes, and then encoding the bytes. Same for decoding -- we'll always know that we have some geo's when decoding the bytes.

otan replied:
i'd rather differentiate for now - especially as we're not (fully) decided on the encoding scheme for geo in protobuf form.

rohany replied:
OK, that makes sense then.

rohany commented on pkg/util/encoding/encoding.go:
```diff
@@ -972,6 +976,63 @@ func decodeTime(b []byte) (r []byte, sec int64, nsec int64, err error) {
+// TODO(otan): this should ideally just be encoded by {SRID,Shape,Raw Points}.
+// EWKB is expensive to encode. However, we don't store this as a PRIMARY KEY
+// (this is needed for GROUP BY only for now), so we ignore it for now.
+func EncodeGeoAscending(b []byte, g *geopb.SpatialObject) ([]byte, error) {
```
As with my comment above, why doesn't this just use encodeBytesAscending?

otan replied:
the marshal / unmarshal abstraction is at the very least helpful in this situation.

rohany commented on pkg/sql/sqlbase/column_type_encoding.go:
```diff
@@ -93,6 +94,16 @@ func EncodeTableKey(b []byte, val tree.Datum, dir encoding.Direction) ([]byte, e
 			return encoding.EncodeStringAscending(b, string(*t)), nil
 		}
 		return encoding.EncodeStringDescending(b, string(*t)), nil
+	case *tree.DGeography:
```
Is it likely that if we add forward indexing for geo types that their encodings will look like this?

otan replied:
yes, if we decide with primary keys it will almost certainly be something like this.

otan commented:
bors r=rohany
